core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -517,47 +526,90 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
           val coordinatorEpoch = epochAndTxnMetadata.coordinatorEpoch
 
           txnMetadata.inLock {
-            if (txnMetadata.producerId != producerId)
+            producerIdCopy = txnMetadata.producerId
+            producerEpochCopy = txnMetadata.producerEpoch
+            val currentTxnMetadataIsAtLeastTransactionsV2 = txnMetadata.clientTransactionVersion >= 2
```

Comment:
I think the logic is mostly correct here (except maybe returning invalid state error where we'd normally return producer fenced error), I wonder if it could be more readable if we had a block of code that calculate 2 Booleans: isValidProducerId and isValidEpoch, each calculation would take the states and versions into consideration.

Reply:
I think that is closer to what I had before -- though I focused on retries as that was what I thought was a non-trivial special case. 

I changed it to the current state since I thought we were trying to keep the current format where we have the various states enumerated.

Reply:
I don't have a strong opinion, if you think this code structure is best, I think it works (I left a comment on the other issue separately).